src/fft_2n.h

```diff
@@ -336,39 +488,135 @@ void Radix2<T>::fft(vec::Buffers<T>& output, vec::Buffers<T>& input)
 template <typename T>
 void Radix2<T>::fft_inv(vec::Buffers<T>& output, vec::Buffers<T>& input)
 {
-    const unsigned size = this->pkt_size;
     const unsigned len = this->n;
     const unsigned input_len = input.get_n();
 
     // 1st reversion of elements of output
     bit_rev_permute(output);
 
+    // copy input to output
+    const std::vector<T*>& i_mem = input.get_mem();
+    const std::vector<T*>& o_mem = output.get_mem();
     unsigned i;
     for (i = 0; i < input_len; ++i) {
-        output.copy(i, input.get(i));
+        memcpy(o_mem[i], i_mem[i], buf_size);
     }
-    for (; i < len; ++i) {
-        output.fill(i, 0);
+
+    unsigned m = len / 2;
+
+    if (input_len < len) {
+        const unsigned input_len_power_2 =
+            arith::get_smallest_power_of_2<unsigned>(input_len);
+        for (; i < input_len_power_2; ++i) {
+            memset(o_mem[i], 0, buf_size);
+        }
+
+        // For Q are zeros only => Q = c * P
+        for (; m >= input_len; m /= 2) {
+            unsigned doubled_m = 2 * m;
+            for (unsigned j = 0; j < m; ++j) {
+                const T r = inv_W->get(j * len / doubled_m);
+                butterfly_gs_step_simple(output, r, j, m, doubled_m);
+            }
+        }
     }
 
-    for (unsigned m = len / 2; m >= 1; m /= 2) {
+    // Next, normal butterlfy GS is performed
+    for (; m >= 1; m /= 2) {
         unsigned doubled_m = 2 * m;
         for (unsigned j = 0; j < m; ++j) {
-            T r = inv_W->get(j * len / doubled_m);
-            for (unsigned i = j; i < len; i += doubled_m) {
-                T* a = output.get(i);
-                T* b = output.get(i + m);
-
-                // perform butterfly operation for Gentleman-Sande FFT algorithm
-                this->gf->butterfly_gs(r, a, b, size);
-            }
+            const T r = inv_W->get(j * len / doubled_m);
+            butterfly_gs_step(output, r, j, m, doubled_m);
         }
     }
 
     // 2nd reversion of elements of output to return its natural order
     bit_rev_permute(output);
 }
 
+// for each pair (P, Q) = (buf[i], buf[i + m]):
+// Q = c * P
+template <typename T>
+void Radix2<T>::butterfly_gs_step_simple(
+    vec::Buffers<T>& buf,
+    T coef,
+    unsigned start,
+    unsigned m,
+    unsigned step)
+{
+    for (int i = start; i < this->n; i += step) {
+        T* a = buf.get(i);
+        T* b = buf.get(i + m);
+        // perform butterfly operation for Cooley-Tukey FFT algorithm
+        for (size_t j = 0; j < this->pkt_size; ++j) {
+            b[j] = this->gf->mul(coef, a[j]);
+        }
+    }
+}
+
+// for each pair (P, Q) = (buf[i], buf[i + m]):
```

Comment:
Can't we merge `butterfly_gs_step` with `butterfly_gs_step_offset` by using a default value for `offset`, i.e

```
template <typename T>
void Radix2<T>::butterfly_gs_step_offset(
    vec::Buffers<T>& buf,
    T coef,
    unsigned start,
    unsigned m,
    unsigned step,
    size_t offset = 0)
```

BTW,  `butterfly_gs_step_offset` seems unused, so it can be removed, no?

Reply:
`butterfly_gs_step_offset` is used in specialized functions: `fft_2n.cpp`.
But I will merge them to avoid duplicated codes.

Reply:
- [ ] Remove duplicated codes

Reply:
Why having

```
template <typename T>
void Radix2<T>::butterfly_gs_step_simple(
    vec::Buffers<T>& buf, T coef, unsigned start, unsigned m, unsigned step
)
{
    butterfly_gs_step_offset_simple(buf, coef, start, m, step, 0);
}

template <typename T>
void Radix2<T>::butterfly_gs_step_offset_simple(
    vec::Buffers<T>& buf, T coef, unsigned start, unsigned m, unsigned step, size_t offset
)
{
    […]
}
```

instead of

```
template <typename T>
void Radix2<T>::butterfly_gs_step_simple(
    vec::Buffers<T>& buf, T coef, unsigned start, unsigned m, unsigned step, size_t offset=0
)
{
    […]
}
```

?

Using default arguments, you don't need two functions: one is enough.

Reply:
The reason is the same for `butterfly_gs_step_offset` as I said above: `butterfly_gs_step_simple` will be specialized, and it will use `butterfly_gs_step_offset_simple` to avoid duplication.
These specialized functions are in `fft_2n.cpp`, you can see it in #254

Reply:
I don't understand, you still can have a single function if you use a default argument.

Did you tried?

I tried (with the following diff, if it can help you) and it works!

```diff
diff --git a/src/fft_2n.cpp b/src/fft_2n.cpp
index dea8e46..efa43ca 100644
--- a/src/fft_2n.cpp
+++ b/src/fft_2n.cpp
@@ -70,18 +70,18 @@ void Radix2<uint16_t>::butterfly_ct_two_layers_step(
         //  ---------
         const uint16_t r1 = W->get(start * this->n / m / 2);
         // first pair
-        butterfly_ct_step_offset(buf, r1, start, m, step, offset);
+        butterfly_ct_step(buf, r1, start, m, step, offset);
         // second pair
-        butterfly_ct_step_offset(buf, r1, start + 2 * m, m, step, offset);
+        butterfly_ct_step(buf, r1, start + 2 * m, m, step, offset);
         //  ---------
         // Second layer
         //  ---------
         // first pair
         const uint16_t r2 = W->get(start * this->n / m / 4);
-        butterfly_ct_step_offset(buf, r2, start, 2 * m, step, offset);
+        butterfly_ct_step(buf, r2, start, 2 * m, step, offset);
         // second pair
         const uint16_t r3 = W->get((start + m) * this->n / m / 4);
-        butterfly_ct_step_offset(buf, r3, start + m, 2 * m, step, offset);
+        butterfly_ct_step(buf, r3, start + m, 2 * m, step, offset);
     }
 }
 
@@ -104,7 +104,7 @@ void Radix2<uint16_t>::butterfly_ct_step(
     // for last elements, perform as non-SIMD method
     if (last_len > 0) {
         size_t offset = vec_len * ratio;
-        butterfly_ct_step_offset(buf, r, start, m, step, offset);
+        butterfly_ct_step(buf, r, start, m, step, offset);
     }
 }
 
@@ -182,18 +182,18 @@ void Radix2<uint32_t>::butterfly_ct_two_layers_step(
         //  ---------
         const uint32_t r1 = W->get(start * this->n / m / 2);
         // first pair
-        butterfly_ct_step_offset(buf, r1, start, m, step, offset);
+        butterfly_ct_step(buf, r1, start, m, step, offset);
         // second pair
-        butterfly_ct_step_offset(buf, r1, start + 2 * m, m, step, offset);
+        butterfly_ct_step(buf, r1, start + 2 * m, m, step, offset);
         //  ---------
         // Second layer
         //  ---------
         // first pair
         const uint32_t r2 = W->get(start * this->n / m / 4);
-        butterfly_ct_step_offset(buf, r2, start, 2 * m, step, offset);
+        butterfly_ct_step(buf, r2, start, 2 * m, step, offset);
         // second pair
         const uint32_t r3 = W->get((start + m) * this->n / m / 4);
-        butterfly_ct_step_offset(buf, r3, start + m, 2 * m, step, offset);
+        butterfly_ct_step(buf, r3, start + m, 2 * m, step, offset);
     }
 }
 
@@ -216,7 +216,7 @@ void Radix2<uint32_t>::butterfly_ct_step(
     // for last elements, perform as non-SIMD method
     if (last_len > 0) {
         size_t offset = vec_len * ratio;
-        butterfly_ct_step_offset(buf, r, start, m, step, offset);
+        butterfly_ct_step(buf, r, start, m, step, offset);
     }
 }
 
diff --git a/src/fft_2n.h b/src/fft_2n.h
index 68f83b1..8829d9e 100644
--- a/src/fft_2n.h
+++ b/src/fft_2n.h
@@ -94,7 +94,8 @@ class Radix2 : public FourierTransform<T> {
         T r,
         unsigned start,
         unsigned m,
-        unsigned step);
+        unsigned step,
+        size_t offset=0);
     void butterfly_ct_two_layers_step(
         vec::Buffers<T>& buf,
         unsigned start,
@@ -113,13 +114,6 @@ class Radix2 : public FourierTransform<T> {
         unsigned step);
 
     // Only used for non-vectorized elements
-    void butterfly_ct_step_offset(
-        vec::Buffers<T>& buf,
-        T coef,
-        unsigned start,
-        unsigned m,
-        unsigned step,
-        size_t offset);
     void butterfly_gs_step_offset(
         vec::Buffers<T>& buf,
         T coef,
@@ -380,29 +374,6 @@ void Radix2<T>::fft(vec::Buffers<T>& output, vec::Buffers<T>& input)
     }
 }
 
-// for each pair (P, Q) = (buf[i], buf[i + m]):
-// P = P + c * Q
-// Q = P - c * Q
-template <typename T>
-void Radix2<T>::butterfly_ct_step(
-    vec::Buffers<T>& buf,
-    T r,
-    unsigned start,
-    unsigned m,
-    unsigned step)
-{
-    for (int i = start; i < this->n; i += step) {
-        T* a = buf.get(i);
-        T* b = buf.get(i + m);
-        // perform butterfly operation for Cooley-Tukey FFT algorithm
-        for (size_t j = 0; j < this->pkt_size; ++j) {
-            T x = this->gf->mul(r, b[j]);
-            b[j] = this->gf->sub(a[j], x);
-            a[j] = this->gf->add(a[j], x);
-        }
-    }
-}
-
 /**
  * Butterly CT on two-layers at a time
  *
@@ -453,7 +424,7 @@ void Radix2<T>::butterfly_ct_two_layers_step(
 }
 
 template <typename T>
-void Radix2<T>::butterfly_ct_step_offset(
+void Radix2<T>::butterfly_ct_step(
     vec::Buffers<T>& buf,
     T coef,
     unsigned start,
@@ -642,7 +613,8 @@ void Radix2<uint16_t>::butterfly_ct_step(
     uint16_t r,
     unsigned start,
     unsigned m,
-    unsigned step);
+    unsigned step,
+    size_t offset=0);
 template <>
 void Radix2<uint16_t>::butterfly_gs_step(
     vec::Buffers<uint16_t>& buf,
@@ -669,7 +641,8 @@ void Radix2<uint32_t>::butterfly_ct_step(
     uint32_t r,
     unsigned start,
     unsigned m,
-    unsigned step);
+    unsigned step,
+    size_t offset=0);
 template <>
 void Radix2<uint32_t>::butterfly_gs_step(
     vec::Buffers<uint32_t>& buf,
```

Reply:
Thank you. I have not tried it :)
In your diff of fft_2n.cpp, I observe infinite loops, e.g.
```
@@ -104,7 +104,7 @@ void Radix2<uint16_t>::butterfly_ct_step(
     // for last elements, perform as non-SIMD method
     if (last_len > 0) {
         size_t offset = vec_len * ratio;
-        butterfly_ct_step_offset(buf, r, start, m, step, offset);
+        butterfly_ct_step(buf, r, start, m, step, offset);
     }
 }
```
I'm wondering if it works. I will check it.

Reply:
Nope, it does not work :(

Reply:
Really? (beware, I made this diff on top of you PR, not on this commit specifically).
Because I ran `make check` and all the test were passing.

Reply:
it's possible to handle the trailing bits in SIMD part, but these functions are very similar to template member functions. Anyway, I'm waiting for your comment :)

Reply:
Yes, you're right :)

I thought about it and it will only move the problem, not solve it.

I can propose you the following:

We define the default slow operation in a well-named method `butterfly_gs_step_simple_slow`

```c++
template <typename T>
void Radix2<T>::butterfly_gs_step_simple_slow(
    vec::Buffers<T>& buf,
    T coef,
    unsigned start,
    unsigned m,
    unsigned step,
    size_t offset=0)
{
    for (int i = start; i < this->n; i += step) {
        T* a = buf.get(i);
        T* b = buf.get(i + m);
        // perform butterfly operation for Cooley-Tukey FFT algorithm
        for (size_t j = offset; j < this->pkt_size; ++j) {
            T x = this->gf->sub(a[j], b[j]);
            a[j] = this->gf->add(a[j], b[j]);
            b[j] = this->gf->mul(coef, x);
        }
    }
}
```

The we have the default template definition that relies on this slow implementation:

```c++
template <typename T>
void Radix2<T>::butterfly_gs_step_simple(
    vec::Buffers<T>& buf, T coef, unsigned start, unsigned m, unsigned step
)
{
    butterfly_gs_step_simple_slow(buf, coef, start, m, step);
}
```

And finally we have the specialized SIMD version that use faster code for the main processing and fallback on the slow implementation for the trailing elements.

```c++
template <>
void Radix2<uint16_t>::butterfly_gs_step_simple(
    vec::Buffers<uint16_t>& buf,
    uint16_t coef,
    unsigned start,
    unsigned m,
    unsigned step)
{
    const unsigned ratio = ALIGN_SIZE / sizeof(uint16_t);
    const size_t len = this->pkt_size;
    const size_t vec_len = len / ratio;
    const size_t last_len = len - vec_len * ratio;

    // perform vector operations
    simd::butterfly_gs_step_simple(buf, coef, start, m, vec_len, card);

    // for last elements, perform as non-SIMD method
    if (last_len > 0) {
        size_t offset = vec_len * ratio;
        butterfly_gs_step_simple_slow(buf, coef, start, m, step, offset);
    }
}
```

What do you think?

Reply:
It's exactly what I did, except the function name and the optional parameter `offset` :)

Reply:
Yup, pretty similar to what you did, but the naming `slow` is more explicit and naming is important :)
(and the default parameter avoid to specify the offset when you don't need it => but that's minor)

Reply:
ok, I will update it :)